chain/checkpointing/sub.go

```diff
@@ -635,62 +867,143 @@ func BuildCheckpointingSub(mctx helpers.MetricsCtx, lc fx.Lifecycle, c *Checkpoi
 	// Get first checkpoint from eudico block 0
 	ts, err := c.api.ChainGetGenesis(ctx)
 	if err != nil {
-		log.Errorf("couldnt get genesis tipset: %v", err)
+		log.Errorf("could not get genesis tipset: %v", err)
 		return
+	} else {
+		log.Infow("Got genesis tipset")
 	}
-	cidBytes := ts.Key().Bytes()// this is the checkpoint (i.e. hash of block)
-	publickey, err := hex.DecodeString(c.cpconfig.PublicKey)
+
+	cidBytes := ts.Key().Bytes()
+	//fmt.Println("cidbytes: ", cidBytes)
+	//fmt.Println("public key before decoding: ", c.cpconfig.PublicKey) // this is the checkpoint (i.e. hash of block)
+	publickey, err := hex.DecodeString(c.cpconfig.PublicKey)          //publickey pre-generated
 	if err != nil {
-		log.Errorf("couldnt decode public key: %v", err)
+		log.Errorf("could not decode public key: %v", err)
 		return
+	} else {
+		log.Infow("Decoded Public key")
 	}
 
-	// Get the last checkpoint from the bitcoin node
-	btccp, err := GetLatestCheckpoint(c.cpconfig.BitcoinHost, publickey, cidBytes)
+	// initialize the kvs
+	r := NewResolver(c.host.ID(), c.ds, c.pubsub)
+	fmt.Println("My id: ",c.host.ID())
+	c.r = r
+	err = c.r.HandleMsgs(ctx)
 	if err != nil {
-		log.Errorf("couldnt decode public key: %v", err)
+		log.Errorf("error initializing cross-msg resolver: %s", err)
 		return
 	}
 
-	// Get the config in minio using the last checkpoint found through Bitcoin.
-	// NOTE: We should be able to get the config regarless of storage (minio, IPFS, KVS,....)
-	cp, err := GetMinersConfig(ctx, c.minioClient, c.cpconfig.MinioBucketName, btccp.cid)
 
-	if cp != "" {
-		// Decode hex checkpoint to bytes
-		cpBytes, err := hex.DecodeString(cp)
-		if err != nil {
-			log.Errorf("couldnt decode checkpoint: %v", err)
-			return
+
+	//eiher send the funding transaction (if needed) or get the latest checkpoint
+	// of the transaction has already been sent
+	if c.taprootConfig != nil {
+		c.pubkey = genCheckpointPublicKeyTaproot(c.taprootConfig.PublicKey, cidBytes)
+
+		// this should be changed such that the public key is updated when eudico is stopped
+		// (so that we can continue the checkpointing without restarting from scratch each time)
+		address, _ := pubkeyToTapprootAddress(c.pubkey)
+		fmt.Println("Address: ", address)
+
+		// only Alice will send the funding transaction for testing purpose
+		if c.host.ID().String() == "12D3KooWMBbLLKTM9Voo89TXLd98w4MjkJUych6QvECptousGtR4" {
+			//start by getting the balance in our wallet (only if sendall is true, i.e. we send all the amount)
+			var value float64
+			if sendall {
+				payload1 := "{\"jsonrpc\": \"1.0\", \"id\":\"wow\", \"method\": \"getbalances\", \"params\": []}"
```

Comment:
I made the comment somewhere else. Using structs and `json.Marshal` for this would make the code way more readable (unless there is a strong reason to do it this way).